src/globus_sdk/transport/retry.py

```diff
@@ -0,0 +1,254 @@
+import random
+import time
+import typing
+
+import requests
+
+from globus_sdk.authorizers import GlobusAuthorizer
+
+
+class RetryContext:
+    """
+    The RetryContext is an object passed to retry checks in order to determine whether
+    or not a request should be retried. The context is constructed after each request,
+    regardless of success or failure.
+
+    If an exception was raised, the context will contain that exception object.
+    Otherwise, the context will contain a response object. Exactly one of ``response``
+    or ``exception`` will be present.
+
+    :param attempt: The request attempt number, starting at 0.
+    :type attempt: int
+    :param response: The response on a successful request
+    :type response: requests.Response
+    :param exception: The error raised when trying to send the request
+    :type exception: Exception
+    :param authorizer: The authorizer object from the client making the request
+    :type authorizer: :class:`GlobusAuthorizer \
+        <globus_sdk.authorizers.GlobusAuthorizer>`
+    """
+
+    def __init__(
+        self,
+        attempt: int,
+        *,
+        retry_state: typing.Dict,
+        response: typing.Optional[requests.Response] = None,
+        exception: typing.Optional[Exception] = None,
+        authorizer: typing.Optional[GlobusAuthorizer] = None,
+    ):
+        # retry attempt number
+        self.attempt = attempt
+        # the response or exception from a request
+        # we expect exactly one of these to be non-null
+        self.response = response
+        self.exception = exception
+        # the authorizer object (which may or may not be able to handle reauth)
+        self.authorizer = authorizer
+        # state which may be accumulated over multiple retry/retry-checker invocations
+        # this is passed forward through all checkers and should be maintained outside
+        # of the context of any singular retry
+        self.retry_state = retry_state
+        # the retry delay or "backoff" before retrying
+        self.backoff: typing.Optional[float] = None
+
+
+def _parse_retry_after(response: requests.Response) -> typing.Optional[int]:
+    val = response.headers.get("Retry-After")
+    if not val:
+        return None
+    try:
+        return int(val)
+    except ValueError:
+        return None
```

Comment:
```suggestion
    try:
        return int(val)
    except TypeError, ValueError:
        return None
```

Reply:
I found that `mypy` yelled at me when I did it this way, because `int(None)` was a possibility (even though TypeError gets caught). Do you not run into the same issue?

I'd prefer to do with a simple try-except if it can pass the type checking.

Reply:
That's weird, I don't get the same error (aside from the example I wrote needing parens on the except). That is `mypy==0.812`. It's a trivial enough thing though if the more verbose version is needed.

Reply:
I'll check again. It's always possible that I did something odd.

Reply:
I'm also using `mypy==0.812`, and this `repro.py` file fails for me:
```python
import typing


def foo(x: typing.Optional[str]):
    try:
        return int(x)
    except (TypeError, ValueError):
        return None
```